compiler_gym/envs/llvm/datasets.py

```diff
@@ -110,12 +158,519 @@
+            cmd,
+            shell=True,
+            stderr=subprocess.STDOUT,
+            stdout=subprocess.PIPE,
+            env=os.environ,
+        )
+
+        try:
+            with Timer() as timer:
+                stdout, _ = process.communicate(timeout=timeout_seconds)
+        except subprocess.TimeoutExpired:
+            process.kill()
+            return BenchmarkExecutionResult(
+                walltime_seconds=timeout_seconds,
+                error=f"Benchmark failed to complete with {timeout_seconds} timeout.",
+            )
+
+        if not ignore_returncode and process.returncode:
+            output = None
+            try:
+                output = stdout.decode("utf-8")
+            except UnicodeDecodeError:
+                pass
+            return BenchmarkExecutionResult(
+                walltime_seconds=timer.time,
+                error=f"Benchmark exited with returncode {process.returncode}. Output: {output}",
+            )
+        return BenchmarkExecutionResult(walltime_seconds=timer.time, output=stdout)
+    finally:
+        bitcode_file.unlink()
+
+
+@lru_cache(maxsize=1024)
+def get_cBench_reference_output(
+    benchmark: str, cmd: str, linkopts: Tuple[str], **hashopts
+) -> str:
+    """Produce a gold-standard benchmark output."""
+    env = gym.make("llvm-v0")
+    try:
+        env.reset(benchmark=benchmark)
+        outcome = _run_cbench_benchmark(
+            env, cmd, num_runs=1, ignore_returncode=False, linkopts=linkopts
+        )
+        if outcome.error:
+            raise OSError(
+                f"Failed to produce reference output for benchmark '{env.benchmark}' using '{cmd}': {outcome.error}"
+            )
+    finally:
+        env.close()
+    return outcome.output
+
+
+@fasteners.interprocess_locked(cache_path(f"cBench-v0-runtime-data.LOCK"))
+def download_cBench_runtime_data() -> bool:
+    """Download and unpack the cBench runtime dataset."""
+    if _CBENCH_DATA.is_dir():
+        return False
+    else:
+        tar_contents = io.BytesIO(
+            download(_CBENCH_DATA_URL, sha256=_CBENCH_DATA_SHA256)
+        )
+        with tarfile.open(fileobj=tar_contents, mode="r:bz2") as tar:
+            _CBENCH_DATA.parent.mkdir(parents=True)
+            tar.extractall(_CBENCH_DATA.parent)
+        assert _CBENCH_DATA.is_dir()
+        return True
+
+
+def make_cBench_validator(
+    cmd: str,
+    linkopts: Tuple[str],
+    os_env: Dict[str, str],
+    num_runs: int = 1,
+    ignore_returncode: bool = False,
+    compare_output: bool = True,
+    input_files: Optional[List[Path]] = None,
+    output_files: Optional[List[Path]] = None,
+    validate_result: Optional[
+        Callable[[BenchmarkExecutionResult], Optional[str]]
+    ] = None,
+    pre_execution_callback: Optional[Callable[[], None]] = None,
+):
+    """Generate a validation callback for a cBench benchmark."""
+    input_files = input_files or []
+    output_files = output_files or []
+
+    def validator(env):
```

Comment:
Would it make sense to ensure that the benchmarks satisfy the LLVM sanitizers like asan, msan etc. both before and after optimization? (if so, can just be a TODO, but those seem valuable - if the original benchmark doesn't satisfy a sanitizer, then optimizations that mess it up aren't necessarily wrong, while introducing san* failures would be bad.)

Reply:
I really like this idea! Yep let's stick a TODO for this. It will take a bit of extra plumbing to get it working (we don't currently ship the clang rt libs) but the value is clear

Reply:
Opened #49 to follow up on this.